explorer/interpreter/type_checker.cpp

```diff
@@ -3443,6 +3504,23 @@ auto TypeChecker::TypeCheckClassDeclaration(
     CARBON_RETURN_IF_ERROR(TypeCheckDeclaration(m, class_scope, class_decl));
     CARBON_RETURN_IF_ERROR(CollectMember(class_decl, m));
   }
+  std::unordered_map<std::string_view, SourceLocation> member_names;
+  const auto check_duplicate_members =
+      [&member_names](Nonnull<Declaration*> m) -> ErrorOr<Success> {
+    if (const auto name = GetName(*m); name.has_value()) {
+      if (auto [_, inserted] = member_names.insert(
+              std::make_pair(name.value(), m->source_loc()));
+          !inserted) {
+        return ProgramError(m->source_loc())
+               << "A member named `" << name.value()
+               << "` already exists in this class or its parents";
+      };
+    }
+    return Success();
+  };
+  CARBON_RETURN_IF_ERROR(
+      VisitAllClassMembers(*class_decl, check_duplicate_members));
```

Comment:
Thanks for the ref to StaticScope.

Supported in C++ yes, but for Carbon the way I read it, we don't want to allow child classes to non-explicitly (override or other) hide a method (making it expressive, not pitfalls, etc.). I'm thinking about this reference: https://github.com/carbon-language/carbon-lang/blob/8e8e8ced6014ae131290a7eb74631cfa58c80c4f/docs/design/classes.md?plain=1#L1906

That said this reference targets member functions specifically. I think I remember something more generic including member variables in the class design, though I'd have to dig a bit deeper (if it does, in fact, exist).

Maybe the intent needs to be clarified too.
Do you have a different interpretation?

Reply:
Thanks, I'd missed that. But note it's under ["Future work"](https://github.com/carbon-language/carbon-lang/blob/8e8e8ced6014ae131290a7eb74631cfa58c80c4f/docs/design/classes.md?plain=1#L1784) -- that means it's been thought about, but not in enough depth to make a decision.

Reply:
Also note the paragraph before the bullets:

> Beyond that, we may want some rules or restrictions about defining methods in a
> derived class with the same name as a base class method without overriding it.
> There are some opportunities to improve on and simplify the C++ story:

I would emphasize "may". While the wording in the bullets is strong, it's wrapped by a tentative statement.

Anyways, I wanted to add that even if we _do_ add this, the approach should change: in addition to StaticScope, mixins need to be considered, and it feels like the current approach is less suitable for access control. Since this is undecided, my thought's really that it may be better to build out the other pieces (particularly access control) and that'd probably help shape how name conflicts would be managed if the leads decide to head in that direction.

Reply:
Done, dropped as not aligned with the spec.